src/serve.rs

```diff
@@ -55,186 +51,221 @@ pub enum Error {
     match result {
-        Ok(res) => {
-            let mut ret_xdr_buf: Vec<u8> = Vec::new();
-            match (
-                strval::to_string(&res),
-                res.write_xdr(&mut Cursor::new(&mut ret_xdr_buf)),
-            ) {
-                (Ok(j), Ok(())) => jsonrpc::Response::Ok(jsonrpc::ResultResponse {
-                    jsonrpc: "2.0".to_string(),
-                    id: id.as_ref().unwrap_or(&jsonrpc::Id::Null).clone(),
-                    result: json!({
-                        "json": j,
-                        "xdr": base64::encode(ret_xdr_buf),
-                    }),
-                }),
-                (Err(err), _) => reply(id, Err(Error::StrVal(err))),
-                (_, Err(err)) => reply(id, Err(Error::Xdr(err))),
-            }
-        }
-        Err(err) => jsonrpc::Response::Err(jsonrpc::ErrorResponse {
+        Ok(res) => jsonrpc::Response::Ok(jsonrpc::ResultResponse {
             jsonrpc: "2.0".to_string(),
             id: id.as_ref().unwrap_or(&jsonrpc::Id::Null).clone(),
-            error: jsonrpc::ErrorResponseError {
-                code: match err {
-                    Error::Serde(_) => -32700,
-                    Error::UnknownMethod => -32601,
-                    _ => -32603,
-                },
-                message: err.to_string(),
-                data: None,
-            },
+            result: res,
         }),
+        Err(err) => {
+            eprintln!("err: {:?}", err);
+            jsonrpc::Response::Err(jsonrpc::ErrorResponse {
+                jsonrpc: "2.0".to_string(),
+                id: id.as_ref().unwrap_or(&jsonrpc::Id::Null).clone(),
+                error: jsonrpc::ErrorResponseError {
+                    code: match err {
+                        Error::Serde(_) => -32700,
+                        Error::UnknownMethod => -32601,
+                        _ => -32603,
+                    },
+                    message: err.to_string(),
+                    data: None,
+                },
+            })
+        }
     }
 }
 
-fn invoke(
-    contract: &String,
-    func: &String,
-    args: &[Value],
-    args_xdr: &[String],
-    ledger_file: &PathBuf,
-) -> Result<ScVal, Error> {
-    let contract_id: [u8; 32] = utils::contract_id_from_str(contract)?;
+fn simulate_transaction(txn_xdr: &str, ledger_file: &PathBuf) -> Result<Value, Error> {
+    // Parse and validate the txn
+    let decoded = base64::decode(txn_xdr).map_err(|_| Error::Xdr(XdrError::Invalid))?;
+    let ops = match TransactionEnvelope::from_xdr(decoded)? {
+        TransactionEnvelope::TxV0(envelope) => envelope.tx.operations,
+        TransactionEnvelope::Tx(envelope) => envelope.tx.operations,
+        TransactionEnvelope::TxFeeBump(envelope) => {
+            let FeeBumpTransactionInnerTx::Tx(tx_envelope) = envelope.tx.inner_tx;
+            tx_envelope.tx.operations
+        }
+    };
+    if ops.len() != 1 {
+        return Err(Error::Xdr(XdrError::Invalid));
```

Comment:
It would be ideal to return to this later and add unique errors for each of these error cases, but I understand we probably don't have time for that now. Can you follow up with that at a later time?

Reply:
The error messages in general need a *bunch* of improvement.

Reply:
Added https://github.com/stellar/soroban-cli/issues/90 to track error message improvements.